server/asset/dcr/dcr.go

```diff
@@ -932,14 +932,30 @@ func (dcr *Backend) getTxOutInfo(ctx context.Context, txHash *chainhash.Hash, vo
 	if err != nil {
 		return nil, nil, nil, fmt.Errorf("failed to decode MsgTx from hex for transaction %s: %w", txHash, err)
 	}
-	txTree := stake.DetermineTxType(msgTx, msgTx.Version == wire.TxVersionTreasury)
-	txOut, pkScript, err := dcr.getUnspentTxOut(ctx, txHash, vout, int8(txTree))
+	tree := determineTxTree(msgTx)
+	txOut, pkScript, err := dcr.getUnspentTxOut(ctx, txHash, vout, tree)
 	if err != nil {
 		return nil, nil, nil, err
 	}
 	return txOut, verboseTx, pkScript, nil
 }
 
+func determineTxTree(msgTx *wire.MsgTx) int8 {
+	// Try with treasury disabled first.
+	txType := stake.DetermineTxType(msgTx, false)
+	if txType != stake.TxTypeRegular {
+		return wire.TxTreeStake
+	}
+
+	// Try with treasury enabled.
+	txType = stake.DetermineTxType(msgTx, true)
+	if txType != stake.TxTypeRegular {
+		return wire.TxTreeStake
+	}
+
+	return wire.TxTreeRegular
+}
```

Comment:
I know this was discussed in matrix at length and this was specifically suggested by @davecgh, but I can't find the discussion now.  But even now I don't understand why it's necessary to check treasury disabled first, then enabled. 

I thought the only check that might fail was `standalone.IsCoinbaseTx`, but that's not applicable to `stake.DetermineTxType` (there is no explicit coinbase check in it, and regular is the default if the stake types do not match).  The only possible concern I saw in `DetermineTxType` was for the new vote tx versions, but the way `stake.CheckSSGenVotes` is written, it doesn't matter if we just say true.

Hence the following for `IsCoinbaseTx` in dcrwallet: https://github.com/decred/dcrwallet/blob/703d6cd3e51379683ee38ac24f4ca984b4bfbe0d/internal/compat/compat.go#L20-L28

```go
// IsEitherCoinBaseTx verifies if a transaction is either a coinbase prior to
// the treasury agenda activation or a coinbse after treasury agenda
// activation.
func IsEitherCoinBaseTx(tx *wire.MsgTx) bool {
	if standalone.IsCoinBaseTx(tx, false) {
		return true
	}
	if standalone.IsCoinBaseTx(tx, true) {
		return true
	}
	return false
}
```

But everywhere else just saying true for `stake.DetermineTxType` because the coinbase check is not done directly and it does saying true for the vote check [does not seem to have the potential to give an incorrect result](https://github.com/decred/dcrd/blob/73d498fe12bf4bfa12e011fbd27b3aa66d7e16cd/blockchain/stake/staketx.go#L1007-L1010).

Maybe I'm wrong.  But we should document the reasoning for all of our treasury handling.

Reply:
Looks like you're right. The conversation that led to this change is above, I resolved it previously but I've unresolved it now so it can be easily spotted.

2 things that I see apply from that conversation:
The only things that changed about the old transactions are:
- Regular tree coinbase transactions
- Stake tree votes

So coinbase checks could produce wrong results but as you noted, `stake.DetermineTxType` does not perform coinbase checks.
It does check votes but again, as you say, the way `stake.CheckSSGenVotes` is written, it doesn't matter if we just say true.

So I'll defer to @davecgh again to re-confirm if it's perpetually safe to just pass true for `isTreasuryEnabled`. Will doc this afterwards.
I do see that `server/asset/dcr` on master currently just passes true:
https://github.com/decred/dcrdex/blob/93b2fe3aaf141f0da46a46b0d0ce584a3b490f51/server/asset/dcr/dcr.go#L493

Reply:
I've gone with `stake.DetermineTxType(msgTx, true)` for now until it becomes clearer that we need to check with both false and true.